libraries/AP_WindVane/AP_WindVane.cpp

```diff
@@ -0,0 +1,320 @@
+
+    switch (_type) {
+        case WindVaneType::WINDVANE_HOME_HEADING:
+            // this is a approximation as we are not considering boat speed and wind speed
+            // do not filter home heading
+            _direction_apparent = wrap_PI(_home_heading - AP::ahrs().yaw);
+            return;
+        case WindVaneType::WINDVANE_PWM_PIN:
+            // this is a approximation as we are not considering boat speed and wind speed
+            // do not filter home heading and pwm type vanes
+            _direction_apparent = wrap_PI(read_PWM_bearing() + _home_heading - AP::ahrs().yaw);
+            return;
+        case WindVaneType::WINDVANE_ANALOG_PIN:
+            apparent_angle_in = read_analog();
+            break;
+        case WindVaneType::WINDVANE_SITL:
+#if CONFIG_HAL_BOARD == HAL_BOARD_SITL
+            apparent_angle_in = read_direction_SITL();
+#endif
+            break;
+    }
+
+    // apply low pass filter if enabled
+    if (is_positive(_vane_filt_hz)) {
+        low_pass_filter_wind_sin.set_cutoff_frequency(_vane_filt_hz);
+        low_pass_filter_wind_cos.set_cutoff_frequency(_vane_filt_hz);
+        // https://en.wikipedia.org/wiki/Mean_of_circular_quantities
+        float filtered_sin = low_pass_filter_wind_sin.apply(sinf(apparent_angle_in), 0.02f);
+        float filtered_cos = low_pass_filter_wind_cos.apply(cosf(apparent_angle_in), 0.02f);
+        _direction_apparent = atan2f(filtered_sin, filtered_cos);
+    } else {
+        _direction_apparent = apparent_angle_in;
+    }
+
+    // make sure between -pi and pi
+    _direction_apparent = wrap_PI(_direction_apparent);
+}
+
+// calibrate windvane
+void AP_WindVane::calibrate()
+{
+    // exit immediately if armed or too soon after start
+    if (hal.util->get_soft_armed()) {
+        return;
+    }
+
+    // return if not calibrating
+    if (_calibration == 0) {
+        return;
+    }
+
+    switch (_type) {
+        case WindVaneType::WINDVANE_HOME_HEADING:
+        case WindVaneType::WINDVANE_PWM_PIN:
+            gcs().send_text(MAV_SEVERITY_INFO, "WindVane: No cal required");
+            _calibration.set_and_save(0);
+            break;
+        case WindVaneType::WINDVANE_ANALOG_PIN:
+            // start calibration
+            if (_cal_start_ms == 0) {
+                _cal_start_ms = AP_HAL::millis();
+                _cal_volt_max = _current_analog_voltage;
+                _cal_volt_min = _current_analog_voltage;
+                gcs().send_text(MAV_SEVERITY_INFO, "WindVane: Analog input calibrating");
+            }
+
+            // record min and max voltage
+            _cal_volt_max = MAX(_cal_volt_max, _current_analog_voltage);
+            _cal_volt_min = MIN(_cal_volt_min, _current_analog_voltage);
+
+            // calibrate for 30 seconds
+            if ((AP_HAL::millis() - _cal_start_ms) > 30000) {
+                // save min and max voltage
+                _analog_volt_max.set_and_save(_cal_volt_max);
```

Comment:
No need to apply a sanity check to ensure that the pins are sufficiently different/not the same?

Reply:
yes, I agree we should add checks the voltages collected during calibration are reasonable.  We will add this and also resolve the to-do item to allow kicking off from a GCS via a mavlink message.

Reply:
I've added this voltage check now (1V difference), clarified the calibration messages to the ground station a bit and allowed the calibration to be kicked off using MAV_CMD_PREFLIGHT_CALIBRATION with the param6 field set to 3 which I think is appropriate because param6 is also used for airspeed (but with param6 set to 2).